File: Java/MergeSort.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.io.File;
import java.util.stream.Collectors;
import java.io.FileNotFoundException;

// Average complexity: O(nlog(n))
// Worst complexity: O(nlog(n))

public class MergeSort {
    public static int[] readDataset(String filePath, int maxSize) {
        Scanner datasetFile;
        try {
            datasetFile = new Scanner(new File(filePath));
        } catch (FileNotFoundException e) {
            System.err.println("Failed to open dataset file");
            return null;
        }

        int[] numbers = new int[maxSize];
        int index = 0;
        while (datasetFile.hasNextInt() && index < maxSize) {
            numbers[index++] = datasetFile.nextInt();
        }
        datasetFile.close();

        return numbers;
    }

    public static void main(String args[]) {
        // Default values
        String filePath = "../Dataset/100k_parc_ordenado.txt";
        int maxSize = 100000;

        // Check if the user provided the input file
        if (args.length >= 1) {
            filePath = args[0];
        }

        if (args.length >= 2) {
            maxSize = Integer.parseInt(args[1]);
        }

        int[] numbers = readDataset(filePath, maxSize);
        if (numbers == null) {
            return;
        }

        long startTime = System.currentTimeMillis();
        mergeSort(numbers);
        long endTime = System.currentTimeMillis();

        long elapsedTime = endTime - startTime;

        System.out.println("Elapsed time: " + elapsedTime + " miliseconds");
    }

    public static void mergeSort(int[] list) {
        if (list.length > 1) {
            int mid = list.length / 2;
            int[] left = Arrays.copyOfRange(list, 0, mid);
            int[] right = Arrays.copyOfRange(list, mid, list.length);

            mergeSort(left);
            mergeSort(right);

            merge(list, left, right);
        }

    }

    public static void merge(int[] mainArray, int[] leftArray, int[] rightArray) {
        int leftIndex = 0;
        int rightIndex = 0;
        int mainIndex = 0;
    
        while (leftIndex < leftArray.length && rightIndex < rightArray.length) {
            if (leftArray[leftIndex] < rightArray[rightIndex]) {
                mainArray[mainIndex] = leftArray[leftIndex];
                leftIndex++;
            } else {
                mainArray[mainIndex] = rightArray[rightIndex];
                rightIndex++;
            }
            mainIndex++;
        }
    
        while (leftIndex < leftArray.length) {
            mainArray[mainIndex] = leftArray[leftIndex];
            leftIndex++;
            mainIndex++;
        }
    
        while (rightIndex < rightArray.length) {
            mainArray[mainIndex] = rightArray[rightIndex];
            rightIndex++;
            mainIndex++;
        }
    }
}

File: C/MergeSort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

// Average complexity: O(nlog(n))
// Worst complexity: O(nlog(n))

void merge(int* array, int* temp, int left, int middle, int right) {
    int leftIndex = left;
    int rightIndex = middle + 1;
    int mergedIndex = left;

    while (leftIndex <= middle && rightIndex <= right) {
        if (array[leftIndex] <= array[rightIndex]) {
            temp[mergedIndex] = array[leftIndex];
            leftIndex++;
        } else {
            temp[mergedIndex] = array[rightIndex];
            rightIndex++;
        }
        mergedIndex++;
    }

    while (leftIndex <= middle) {
        temp[mergedIndex] = array[leftIndex];
        leftIndex++;
        mergedIndex++;
    }

    while (rightIndex <= right) {
        temp[mergedIndex] = array[rightIndex];
        rightIndex++;
        mergedIndex++;
    }

    for (int i = left; i <= right; i++) {
        array[i] = temp[i];
    }
}

void mergeSort(int* array, int* temp, int left, int right) {
    if (left < right) {
        int middle = left + (right - left) / 2;

        mergeSort(array, temp, left, middle);
        mergeSort(array, temp, middle + 1, right);

        merge(array, temp, left, middle, right);
    }
}

void writeResult(clock_t executionTime, char* fileName, char* resultPath) {
    FILE *resultsFile;
    resultsFile = fopen(resultPath, "a");
    if (resultsFile == NULL) {
        perror("Erro ao abrir o resultsFile");
        return;
    }

    char *name = strrchr(fileName, '/');
    //fprintf(resultsFile, "C - MergeSort - File: %s\n", name);
    fprintf(resultsFile, "Execution time: %lf ms\n", ((double)executionTime) / ((CLOCKS_PER_SEC / 1000)));
    fclose(resultsFile);
}

int main(int argc, char *argv[]) {
    // Default values
    char filePath[256] = "../../Dataset/1000k_parc_ordenado.txt";
    char resultPath[256] = "../../Results/MergeSort.txt";
    size_t maxSize = 1000000;

    // Check if the user provided the input file
    if (argc >= 2) {
        strncpy(filePath, argv[1], sizeof(filePath) - 1);
        filePath[sizeof(filePath) - 1] = '\0'; // Ensure null-termination
    }

    if (argc >= 3) {
        strncpy(resultPath, argv[2], sizeof(resultPath) - 1);
        resultPath[sizeof(resultPath) - 1] = '\0'; // Ensure null-termination
    }

    if (argc >= 4) {
        maxSize = (size_t)atoi(argv[3]);
    }

    FILE *datasetFile = fopen(filePath, "r");

    if (!datasetFile) {
        fprintf(stderr, "Failed to open dataset file\n");
        return 1;
    }

    int number;
    clock_t executionTime;
    int *numbers = malloc(maxSize * sizeof(int));

    size_t count = 0;
    while (fscanf(datasetFile, "%d", &number) == 1 && count < maxSize) {
        numbers[count++] = number;
    }
    fclose(datasetFile);

    executionTime = clock();
    int *temp = malloc(maxSize * sizeof(int)); // Allocate temporary array
    mergeSort(numbers, temp, 0, maxSize - 1);
    executionTime = clock() - executionTime;
    
    printf("Execution time: %lf ms\n", ((double)executionTime) / ((CLOCKS_PER_SEC / 1000)));

    writeResult(executionTime, filePath, resultPath);

    free(numbers);
    free(temp);
    return 0;
}